### PR TITLE
LED with isolator heuristic

### PR DESCRIPTION
**Pull request by edongashi**
No unknown column as of yet. Still trying to figure out an efficient approach...

### CODE REVIEW

File: anonymizer/OpenDiffix.Service/Utils.fs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    match findAggregator DiffixLowCount aggregationContext.Aggregators with
+    | Some index -> index
+    | None -> failwith "Cannot find required DiffixLowCount aggregator"
+
+  let diffixCountIndex (aggregationContext: AggregationContext) =
+    match findAggregator DiffixCount aggregationContext.Aggregators with
+    | Some index -> index
+    | None -> failwith "Cannot find required DiffixCount aggregator"
+
+  let safeIter fn (seq: 'a seq) : 'a array =
```
Comment (pdobacz) [inline anonymizer/OpenDiffix.Service/Utils.fs:35]: please add some comment and/or a readable function name: what does this do?

File: anonymizer/OpenDiffix.Service/Led.fs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+  for i in aggIndexes do
+    targetAggregators.[i].Merge(sourceAggregators.[i])
+
+type ColumnState = { Values: Dictionary<Value, int>; mutable TopValues: (Value * int) list }
+
+[<Literal>]
+let TOP_VALUES = 3
+
+let rec private removeValue index value list =
```
Comment (pdobacz) [inline anonymizer/OpenDiffix.Service/Led.fs:44]: what does this and the next function do exactly?

File: anonymizer/OpenDiffix.Service/Led.fs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+let private isolatingColumns (aggregationContext: AggregationContext) (buckets: Bucket seq) =
+  let countsByColumn =
+    Array.init aggregationContext.GroupingLabels.Length (fun _ -> { Values = Dictionary(); TopValues = [] })
+
+  let mutable totalRows = 0
+
+  let buckets =
+    buckets
+    |> Utils.safeIter (fun bucket ->
+      let count = bucket.RowCount
```
Comment (pdobacz) [inline anonymizer/OpenDiffix.Service/Led.fs:71]: if possible, can we refactor this into a well-named function to document this step? Hm, actually this comment applies to all the lambdas here. If we could extract into named functions, it would be easier to grasp the overall flow of the algo

File: anonymizer/OpenDiffix.Service/Led.fs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  Logger.debug $"[LED] Buckets with isolating column(s): {bucketsIsolatingColumn}"
+  Logger.debug $"[LED] Merged buckets: {bucketsMerged}"
+  Logger.debug $"[LED] Merge distribution: {mergeHistogramStr}"
+  Logger.debug $"[LED] Total comparisons: {comparisons}"
+  Logger.debug $"[LED] Runtime: {stopwatch.Elapsed}"
+
+  buckets :> Bucket seq
+
+let hook (aggregationContext: AggregationContext) (buckets: Bucket seq) =
+  if aggregationContext.GroupingLabels.Length <= 2 then
```
Comment (pdobacz) [inline anonymizer/OpenDiffix.Service/Led.fs:217]: explain the motivation behind this short circuit

File: anonymizer/OpenDiffix.Service/Led.fs [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+  let isolatingColumnsStr = isolatingColumns |> Seq.map string |> String.join ", "
+
+  let mergeHistogramStr =
+    mergeHistogram
+    |> Seq.sortBy (fun pair -> pair.Key)
+    |> Seq.map (fun pair -> $"({pair.Key}, {pair.Value})")
+    |> Seq.append [ $"(*, {mergeHistogram |> Seq.sumBy (fun pair -> pair.Value)})" ]
+    |> String.join " "
+
+  Logger.debug $"[LED] Total buckets: {buckets.Length}"
```
Comment (pdobacz) [inline anonymizer/OpenDiffix.Service/Led.fs:204]: maybe it would be worth to extract everything diagnostic-related to a `Led.Diagnostics` module to make it stand out visually and make following the algo easier?

**Review by pdobacz — COMMENTED** [review]
I find it a bit hard to follow. Could you please split the code up into smaller functions with names and/or some brief docs explaining what they accomplish?

A 1-2 sentence overview of the idea of "LED with  isolator heuristic" would help, just to distill the takeaways of the discussions in one place.

**Comment by edongashi** [comment]
> Could you please split the code up into smaller functions with names and/or some brief docs explaining what they accomplish?

Yes, that is the goal. But this design still needs to be verified first, hence the extensive diagnostics. After getting unknown columns working I'll refactor to something better documented and more readable.

File: anonymizer/OpenDiffix.Service/Led.fs (hunk 3 above)
Comment (edongashi) [reply anonymizer/OpenDiffix.Service/Led.fs:71]: Should we do local functions or named functions with more verbose params?

File: anonymizer/OpenDiffix.Service/Led.fs (hunk 5 above)
Comment (edongashi) [reply anonymizer/OpenDiffix.Service/Led.fs:204]: Good idea! A class or something which holds this state could work well

File: anonymizer/OpenDiffix.Service/Led.fs (hunk 3 above)
Comment (pdobacz) [reply anonymizer/OpenDiffix.Service/Led.fs:71]: no idea :/, I always ask myself the same question, and I can hardly make up my mind in most cases :joy:. I guess it depends on how many variables from the outside scope are used and what is the overall layout and role of the function.

**Comment by pdobacz** [comment]
> > Could you please split the code up into smaller functions with names and/or some brief docs explaining what they accomplish?
> 
> Yes, that is the goal. But this design still needs to be verified first, hence the extensive diagnostics. After getting unknown columns working I'll refactor to something better documented and more readable.

fair enough if that's WIP, but it makes reviewing hard and contributions impossible. It doesn't need to be perfect and elaborate refactoring. If you don't want to do it now, just plop in comments before the inline bodies, a 1-sentence summary about what each code block does. Then comments are easily removed, when the final design crystalizes. WDYT?

**Comment by edongashi** [comment]
I started refactoring some parts, and will be explicit for those that are WIP.

**Review by pdobacz — APPROVED** [review]
much easier to get through now, thank you
